BinaryTree.java

```diff
@@ -0,0 +1,148 @@
+    }
+}
+
+public class BinaryTree {
+    private final int info;
+    private final BinaryTree left;
+    private final BinaryTree right;
+
+    public BinaryTree(int info, BinaryTree left, BinaryTree right) {
+        this.info = info;
+        this.left = left;
+        this.right = right;
+    }
+
+    public int getInfo() {
+        return info;
+    }
+
+    public String printAncestors(int key) throws InexistentKeyException{
+        LinkedList<BinaryTree> ancestors =  this.findAncestors(key);
+        if (ancestors.size() == 0) {
+            return "";
+        }
+
+        /* initialize String Builder with probably enough space - 6 chars per key (including space separator) */
+        StringBuilder sb = new StringBuilder(6 * ancestors.size());
+
+        for (BinaryTree crtNode : ancestors) {
+            sb.append(crtNode.getInfo());
+            sb.append(" ");
+        }
+        sb.deleteCharAt(sb.length() - 1);
+
+        return sb.toString();
+    }
+
+    private LinkedList<BinaryTree> findAncestors(int key) throws InexistentKeyException {
+        LinkedList<BinaryTree> ancestors = new LinkedList<BinaryTree>();
+
+        if (this.info == key) {
+            return ancestors;
+        }
+
+        LinkedList leftAncestors = null;
+        LinkedList rightAncestors = null;
+
+        /* look for key left and right, continue tree exploration if key not found */
+        if (this.left != null) {
+            try {
+                leftAncestors = this.left.findAncestors(key);
+                leftAncestors.add(this);
+                return leftAncestors;
+            } catch (InexistentKeyException e) {
+
+            }
+        }
+        if (this.right != null) {
+            try {
+                rightAncestors = this.right.findAncestors(key);
+                rightAncestors.add(this);
+                return rightAncestors;
+            } catch (InexistentKeyException e) {
+            }
+        }
+
+        throw new InexistentKeyException();
+    }
+
+    public BinaryTree lowestCommonAncestor(int key1, int key2) throws InexistentKeyException {
+
+        /* find ancestors of key 1, take each of them and see if the tree rooted there contains key2 */
+        LinkedList<BinaryTree> ancestorsKey1 = findAncestors(key1);
+
+        if(ancestorsKey1.size() == 0) {
+            if (this.containsKey(key2)) {
+                return this;
+            }
+            throw new InexistentKeyException();
+        }
+
+        for (int i = 0 ; i < ancestorsKey1.size() ; i++) {
+
+            BinaryTree intermedNode = ancestorsKey1.get(i);
+            int prevAncestor;
+            if (i > 0) {
+                prevAncestor = ancestorsKey1.get(i-1).getInfo();
+            }
+            else {
+                prevAncestor = key1;
+            }
+
+            if (intermedNode.getInfo() == key2) {
+                return intermedNode;
+            }
+
+
+            if (intermedNode.left != null) {
```

Comment:
This logic with multi-level ifs is quite complex. Could it be simplified? What if you had list of ancestors of both of key1 and of key2? Whould that help?

Side note: solution with lists of ancestors is technically not allowed, since it copies parts of the tree to a separate data structure, which was forbidden (in a slightly convoluted way) by assignment specification. I missed this limitation myself and shouldn't have asked you to do it, and I apologize for doing so, but since we've already taken this route, let's get it right :)

In the future please feel free to disagree if I ask you to do something that seems to be against requirements.